### PR TITLE
Resolution Fix, Boundaries, and Depletion

### PR DESCRIPTION
**Pull request by nicogno**
- Fix the previous error for the resolution
- Introduce Neumann and Dirichlet Boundaries
- Add functionality for depleting substances with other substances

<details>
<summary><b>The BDM PR checklist</b></summary>

First time contributor? Make sure to review our contributing guidelines:
* [Code Quality](https://biodynamo.org/docs/devguide/code_quality/)
* [C++ Style Guide](https://biodynamo.org/docs/devguide/contribute/)

- [x] **Basics**
  - [x] Code builds locally
  - [x] All unit tests pass locally
  - [x] Code passes CI
  - [x] No IDE artifacts such as `.DS_Store` files (only relevant source code)
  - [x] If for some reason your code produces artifacts, make sure to update the
        `.gitignore` file
  - [ ] No new compiler warnings introduced (treat them as an error)

- [x] **New files**
  - [x] License header
  - [x] Include guard
  - [x] BDM namespace

- [x] **New Code**
  - [ ] All new public, protected, and private classes, methods, data members, 
        and functions have full Doxygen-style documentation in source comments.
        Documentation should include descriptions of member data, function
        arguments and return values, template parameters, and prerequisites for
        calling new functions.
  - [x] Pointer arguments and return values must specify whether ownership is
        being transferred or lent with the call.
  - [x] If a method is not thead safe, mark it as such. Generally test if
        threading affects your code (e.g. check if it yields the same results in
        parallel regions with and without `OMP_NUM_THREADS=1`)
  - [x] Coverage via CI, i.e. new code needs to be tested
  - [x] Unit tests check function to a sufficient degree (quality of test needs
        to be verified by reviewer)
  - [x] Code formatted with clang format (see `repository-check` CI)
  - [x] Follows the 
        [C++ style guide](https://google.github.io/styleguide/cppguide.html)
        (must be checked by reviewer)
  - [x] Address all (or at least the vast majority) of the code smells flagged
        by `SonarCloud`
  - [x] New code should not break backward compatibility

- [ ] **New dependencies**
  - [ ] New dependencies should ideally be optional, e.g. can be turned off with
        `-Doption=off`
  - [ ] New dependencies must be compatible with the Apache2.0 license
  - [ ] Document in new dependencies on the website

- [x] **New feature**
  - [x] Update documentation for website (if significant feature)
  - [ ] Consider adding notebook or example showing how to use the new feature

- [ ] **New example**
  - [ ] Make sure that the example is checked in the System CIs
  - [ ] Verify that example appears on the website afterward and works
        (if the example showcases a new feature, the GitPod container may have
        to be updated)
  
Merge instructions: Squash merge PRs and remove individual commit messages from
the merge commit. Make sure to give credit to all contributors via `Co-authored by`.
Remember to delete the branch after merging.

</details>

### CODE REVIEW

**Comment by TobiasDuswald** [comment]
This PR will contain 3 major contributions. The tasks are sorted by contributions

1. **Resolution**

- [ ] resturcture functions to avoide repetitive code
- [ ] add `assert`
- [ ] change missleading discription in comment

2. **Depletion**

- [ ] Don't change `DiffusionGrid`
- [ ] Don't change `EulerGrid`
- [ ] Add `DepletedEulerGrid` subclass (should hanlde multiple substances depleting it)
- [ ] Use `OpenMP` and `SIMD` instructions to optimize update

3. **Boundary Conditions**

- [ ] Introduce NeumandBoundary
- [ ] Introduce DirichletBoundary
- [ ] Deprecate previous boundaries
- [ ] Document boundaries
- [ ] Ideally find analytic solutions to Neuman and Dirichlet Problem and come up with tests

**Comment by TobiasDuswald** [comment]
ToDo: Benchmark to so if `BoundaryCondition::Evaluate(size_t,size_t,size_t,size_t)` introduces a big overhead compared to simply using a constant.

File: src/core/diffusion/euler_grid.cc [hunk 1]
```diff
@@ -143,4 +154,140 @@ void EulerGrid::DiffuseWithOpenEdge(real_t dt) {
...
+        size_t b{0};
+        size_t t{0};
+        c = x + y * nx + z * nx * ny;
+#pragma omp simd
+        for (x = 0; x < nx; x++) {
+          if (x == 0 || x == (nx - 1) || y == 0 || y == (ny - 1) || z == 0 ||
+              z == (nz - 1)) {
+            c2_[c] = boundary_condition_->Evaluate(x, y, z, sim_time);
+            ++c;
+            continue;
```
Comment (TobiasDuswald) [inline src/core/diffusion/euler_grid.cc:189]: I think I'd prefer
```
if ("box == boundary"){
  // treat boundary
} else {
  // Standard heat equation stencil
}
c++;
```
which would quite clearly differentiate between boundary and regular updates

File: src/core/diffusion/euler_grid.cc [hunk 2]
```diff
@@ -143,4 +154,140 @@ void EulerGrid::DiffuseWithOpenEdge(real_t dt) {
...
+            k_comp = -box_length_ *
+                         boundary_condition_->Evaluate(x, y, z, sim_time) -
+                     c1_[c] + c1_[t];
+          } else if (z == (nz - 1)) {
+            k_comp = -box_length_ *
+                         boundary_condition_->Evaluate(x, y, z, sim_time) -
+                     c1_[c] + c1_[b];
+          } else {
+            k_comp = c1_[b] - 2 * c1_[c] + c1_[t];
+          }
```
Comment (TobiasDuswald) [inline src/core/diffusion/euler_grid.cc:280]: Would that be more readable if we went for 

```
          real_t left{c1_[c - 1]};
          real_t right{c1_[c + 1]};
          real_t bottom{c1_[b]};
          real_t top{c1_[t]};
          real_t north{c1_[n]};
          real_t south{c1_[s]};
          real_t center_factor{6.0};

          if (x == 0) {
            left =
                -box_length_ * boundary_condition_->Evaluate(x, y, z, sim_time);
            center_factor -= 1.0;
          } else if (x == (nx - 1)) {
            right =
                -box_length_ * boundary_condition_->Evaluate(x, y, z, sim_time);
            center_factor -= 1.0;
          }

          if (y == 0) {
            north =
                -box_length_ * boundary_condition_->Evaluate(x, y, z, sim_time);
            center_factor -= 1.0;
          } else if (y == (ny - 1)) {
            south =
                -box_length_ * boundary_condition_->Evaluate(x, y, z, sim_time);
            center_factor -= 1.0;
          }

          if (z == 0) {
            bottom =
                -box_length_ * boundary_condition_->Evaluate(x, y, z, sim_time);
            center_factor -= 1.0;
          } else if (z == (nz - 1)) {
            top =
                -box_length_ * boundary_condition_->Evaluate(x, y, z, sim_time);
            center_factor -= 1.0;
          }

          c2_[c] = c1_[c] * (1 - mu_ * dt) +
                   (d * dt * ibl2) * (left + right + south + north + top +
                                      bottom - center_factor * c1_[c]);

          ++c;
```
?

File: src/core/diffusion/runge_kutta_grid.h [hunk 3]
```diff
@@ -54,6 +54,8 @@ class RungeKuttaGrid : public DiffusionGrid {
 
   void DiffuseWithClosedEdge(real_t dt) override;
   void DiffuseWithOpenEdge(real_t dt) override;
+  void DiffuseWithDirichlet(real_t dt) override;
```
Comment (TobiasDuswald) [inline src/core/diffusion/runge_kutta_grid.h:57]: Add `/// Not implemented`

File: doc/user_guide/diffusion.md [hunk 4]
```diff
@@ -140,6 +140,36 @@ Then click the Play button at the top of the screen to run the simulation visual
...
+The `ModelInitializer` conveniently wraps the access to the `ResourceManager`
+and sets the boundaries. You can also set the boundaries directly by calling
+member functions of the `DiffusionGrid` (see API documentation).
+
+If you want to implement more sophisticated boundaries (e.g. with spatial 
+dependence), you may derive classes from `BoundaryCondition` and implement its
+member `BoundaryCondition::Evaluate(size_t,size_t,size_t,real_t)` accordingly.
+If your application required different equations, different domains, different
+numerical schemes, please consult the API for `Continuum`, `ScalarField`, and
+`VectorField` to see how to interface with continuum models with the BioDyanMo
```
[the remark this answers is not in the document]
Comment (TobiasDuswald) [reply doc/user_guide/diffusion.md:170]: ✅

**Review by TobiasDuswald — APPROVED** [review]
Looks good now!

**Comment by TobiasDuswald** [comment]
Thanks @nicogno , that's a great (and absolutely necessary) feature! Great Job!

**Comment by nicogno** [comment]
@TobiasDuswald thank you for your huge help!

**Pull request merged**